File: uts-review/Kompleks/Kompleks.cpp
#include <iostream>
#include "Kompleks.hpp"
using namespace std;

Kompleks::Kompleks() : real(0), imag(0) {};

Kompleks::Kompleks(double r, double i) : real(r), imag(i) {};

Kompleks::Kompleks(const Kompleks &other)
{
    this->real = other.real;
    this->imag = other.imag;
}

double Kompleks::getReal() const
{
    return real;
}

double Kompleks::getImag() const
{
    return imag;
}

double Kompleks::setReal(double r)
{
    return this->real = r;
}

double Kompleks::setImag(double i)
{
    return this->imag = i;
}

Kompleks Kompleks::operator+(const Kompleks &other) const
{
    double tempReal = this->real + other.real;
    double tempImag = this->imag + other.imag;
    return Kompleks(tempReal, tempImag);
}

Kompleks Kompleks::operator-(const Kompleks &other) const
{
    double tempReal = this->real - other.real;
    double tempImag = this->imag - other.imag;

    return Kompleks(tempReal, tempImag);
}

Kompleks Kompleks::operator*(const Kompleks &other) const
{
    // double tempX = this->getReal() * other.real + imag * other.imag;
    // double tempY = this->getReal() * other.imag + imag * other.real;
    double tempX = this->real * other.real + this->imag * other.imag * -1;
    // cout<<this->imag;
    // cout<<other.imag;
    double tempY = this->real * other.imag + this->imag * other.real;
    return Kompleks(tempX, tempY);
}

bool Kompleks::operator==(const Kompleks &other) const
{
    if (this->real == other.real && this->imag == other.imag)
    {
        return true;
    }

    return false;
}

Kompleks &Kompleks::operator=(const Kompleks &other)
{
    if (this != &other)
    {
        this->imag = other.imag;
        this->real = other.real;
    }
    return *this;
}

ostream &operator<<(ostream &os, const Kompleks &c)
{
    if (c.imag == 0)
    {
        os << c.getReal();
    }
    else if (c.getImag() > 0)
    {
        os << c.getReal() << " + " << c.getImag() << "i";
    }
    else if (c.getImag() < 0)
    {
        os << c.getReal() << " - " << c.getImag() * -1 << "i";
    }
    return os;
}

// int main()
// {
//     // Kompleks a;
//     // cout << a;
//     // Kompleks c(3, 4);
//     // // cout << c.getReal();
//     // // cout << c.getImag();
//     // // cout << c;
//     // Kompleks b(3, 4);
//     // // cout << c+b;
//     // Kompleks a=b;
//     // cout << a;

//     Kompleks z(-1, 0);
//     // Kompleks c(3, 4);
//     // cout << c*b;
//     z.setImag(-3);
//     cout << z;
// };

File: 24/tutor4/Kompleks.java

public class Kompleks {
    private int real;
    private int imag;
    private static int n_kompleks;

    // public Kompleks() {
    //     this.real = 0;
    //     this.imag = 0;
    // }
    public Kompleks() {
        this(0,0);
    }

    public Kompleks(int real, int imag) {
        this.real = real;
        this.imag = imag;
    }

    public int getReal() {
        return real;
    }

    public int getImag() {
        return imag;
    }

    public void setReal(int r) {
        this.real = r;
    }

    public void setImaginer(int i) {
        this.imag = i;
    }

    public Kompleks plus(Kompleks b) {
        return new Kompleks(this.real + b.getReal(), this.imag + b.getImag());
    }

    
    public Kompleks minus(Kompleks b) {
        return new Kompleks(this.real - b.getReal(), this.imag - b.getImag());
    }

    // (2 + i)(3 + 2i) = 6 + 4i + 3i + 2i^2
    public Kompleks multiply(Kompleks b) {
        return new Kompleks((this.real * b.getReal() - this.imag * b.getImag()),  (this.imag * b.getReal() + this.real * b.getImag() ));
    }

    public Kompleks multiply (int n){
        return new Kompleks(real * n, imag * n);
    }

    public static int countKompleksInstance() {
        return Kompleks.n_kompleks;
    }

    public void print(){
        if (imag < 0){
            if (real == 0) System.out.println(imag + "i");
            else System.out.println(real + Integer.toString(imag) + "i");
        } else if (imag > 0) {
            if (real == 0) System.out.println(imag + "i");
            else System.out.println(real + "+" + imag + "i");
        } else {
            System.out.println(real);
        }
    }
}